applications/OptimizationApplication/python_scripts/filtering/filter.py

```diff
@@ -0,0 +1,146 @@
+        not exist until the control construction is finished.
+
+        Args:
+            component_data_view (ComponentDataView): Component data view of the control.
+        """
+        self.__component_data_view = component_data_view
+
+    def GetComponentDataView(self) -> ComponentDataView:
+        """Retrieve the component data view.
+
+        This method retrieves the component data view of the control, which can be used
+        to store data and fields.
+
+        Raises:
+            RuntimeError: If SetComponentDataView is not called first before calling this.
+
+        Returns:
+            ComponentDataView: Component data view of the control.
+        """
+        if self.__component_data_view is None:
+            raise RuntimeError("Please use SetComponentDataView first.")
+        return self.__component_data_view
+
+    @abc.abstractmethod
+    def Initialize(self) -> None:
+        """Initializes the filter.
+
+        This method initializes the filter. This is only called once in the whole optimization process.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def Check(self) -> None:
+        """Checks the filter.
+
+        This method checks the filter. This is only called once in the whole optimization process after calling the initialize.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def Finalize(self) -> None:
+        """Finalizes the filter.
+
+        This method checks the filter. This is only called once in the whole optimization process at the end.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def FilterField(self, unfiltered_field: ContainerExpressionTypes) -> ContainerExpressionTypes:
+        """Filter the input unfiltered_field.
+
+        This method filters the passed input field. Unfiltered field is assumed to be a
+        non-integrated field over entity domains.
+
+        Args:
+            unfiltered_field (ContainerExpressionTypes): Input field
+
+        Returns:
+            ContainerExpressionTypes: Filtered output field.
+        """
+        pass
+
+    @abc.abstractmethod
+    def FilterIntegratedField(self, unfiltered_integrated_field: ContainerExpressionTypes) -> ContainerExpressionTypes:
+        """Filter the input integrated field.
+
+        This method filters the passed integrated input field. The unfiltered integrated field is
+        assumed to be already integrated over entity domains.
+
+        Args:
+            unfiltered_integrated_field (ContainerExpressionTypes): Unfiltered integrated field.
+
+        Returns:
+            ContainerExpressionTypes: Filtered output field.
+        """
+        pass
+
+    @abc.abstractmethod
+    def Update(self) -> None:
+        """Updates the filter.
+
+        This method is used to update the filter state. This will be called by the control, when control changes
+        the optimization domain.
+
+        """
+        pass
```

Comment:
I would also add SetFilterRadius method which is also usefull

Reply:
Hmm... Then the question is, who knows about the filtering radius. Should the control needs to know about filtering radius? If so I agree on putting this method.

Another point is, If we are to put this method, then we are making slowly tight ties between `Control` and `Filter`, which is not so nice for modularity. Modularity needs loosely coupling as much as possible.

Reply:
I see it exactly otherway round, for modularity reasons, we filter class should only apply filtering operations for the given settings like filter radius. If a control uses filter then it has a filtering block settings which provides the filter radius and settings.

Reply:
In the case of `constant` Filter kerenel, there is no need of `filter_radius` because it does not use it. So, `filter_radius` is not universal for all the filters. Same in the case with some advanced `filter_radius` calculation methods such as adaptive filter radius methods. 

In the other hand, if we are to put this in the control, then every control needs to get updated whenever someone implement some different filtering radius calculation method, The most modular and flexible way is to allow more freedom, not to restrict it by tightly coupling.

And, if the filter radius falls within the "filter_settings", then it should be read by the filter. Because, someone can call this settings with a different name. I don't like control reading the settings of the filter and passing that setting via a interface method.

Reply:
Ok good, I am fine with it.